translation(CSS): css/css_flexible_box_layout/basic_concepts_of_flexbox/index.md (#175)
webdoky/content: files/uk/web/css/css_flexible_box_layout/basic_concepts_of_flexbox/index.md

### Головна вісь

Головна вісь визначається властивістю `flex-direction`, котра має чотири можливі значення:

- `row`
- `row-reverse`
- `column`
- `column-reverse`

Якщо обрати `row` або `row-reverse`, то головна вісь буде напрямлена горизонтально у **рядковому напрямку**.

![Якщо flex-direction має значення row, то головна вісь напрямлена горизонтально у рядковому напрямку.](basics1.png)

Якщо обрати `column` або `column-reverse`, то головна вісь буде напрямлена вертикально – у **блоковому напрямку**.

![Якщо flex-direction має значення column, то головна вісь напрямлена вертикально у блоковому напрямку.](basics2.png)

### Поперечна вісь

Поперечна вісь напрямлена перпендикулярно до головної, тож якщо `flex-direction` (головна вісь) має значення `row` або `row-reverse`, то поперечна вісь напрямлена вертикально.

![Якщо flex-direction має значення row, то поперечна вісь напрямлена у блоковому напрямку.](basics3.png)

Якщо головна вісь – `column` або `column-reverse`, то поперечна вісь напрямлена горизонтально.

![Якщо flex-direction має значення column, то поперечна вісь напрямлена у рядковому напрямку.](basics4.png)

> Розуміння значення кожної з осей важливе для початку розгляду шикування та вирівнювання гнучких елементів; флексбокс пропонує властивості, що шикують та вирівнюють вміст за однією віссю чи інакшою.

Reviewer: ```suggestion
Розуміння значення кожної з осей важливе для початку розгляду шикування та вирівнювання гнучких елементів; флексбокс пропонує властивості, що шикують та вирівнюють вміст відносно тієї чи іншої осі.
```

Author: ```suggestion
Розуміння значення кожної з осей важливе для початку розгляду шикування та вирівнювання гнучких елементів; флексбокс пропонує властивості, що шикують та вирівнюють вміст відносно тієї чи тієї осі.
```
https://web.posibnyky.vntu.edu.ua/icgn/radomska_ukr_mova/9.htm

## Початкова та кінцева лінії

> Іще одна життєво важлива для розуміння тема – те, як флексбокс не робить припущень щодо режиму письма документа. У минулому CSS сильно схилялась в бік режимів письма горизонтальних та зліва направо. Сучасні способи розкладки охоплюють низку режимів письма, і більше не робиться припущень про те, що рядок тексту почнеться у лівому верхньому куті документа і прямуватиме в бік правої руки, а наступні рядки з‘являться знизу, один за одним.

Reviewer: ```suggestion
Іще одна життєво важлива для розуміння тема – те, як флексбокс не робить припущень щодо режиму письма документа. У минулому CSS сильно схилялась в бік горизонтальних режимів письма з написанням тексту зліва направо. Сучасні способи розкладки охоплюють низку режимів письма, і більше не робиться припущень про те, що рядок тексту почнеться у лівому верхньому куті документа і прямуватиме в бік правої руки, а наступні рядки з‘являться знизу, один за одним.
```

У живому прикладі нижче видно, як це працює. Спробуйте редагувати елементи або додати елементи, щоб перевірити усталену поведінку флексбокса.

{{EmbedGHLiveSample("css-examples/flexbox/basics/the-flex-container.html", '100%', 480)}}

### Зміна flex-direction

Властивість гнучкого контейнера {{cssxref("flex-direction")}} дає змогу змінити напрям виводу гнучких елементів. Встановлення `flex-direction: row-reverse` змусить елементи надалі виводитись в ряд, але змінить місцями початок та кінець ряду.

Якщо змінити значення `flex-direction` на `column`, то головна вісь перемкнеться, і елементи будуть виведені у стовпчик. Якщо задати `column-reverse`, то початок та кінець знову поміняються місцями.

У живому прикладі нижче `flex-direction` має значення `row-reverse`. Спробуйте інші значення: `row`, `column` і `column-reverse`, – і погляньте, що станеться із вмістом.

{{EmbedGHLiveSample("css-examples/flexbox/basics/flex-direction.html", '100%', 350)}}

> ## Багаторядові гнучкі контейнери із flex-wrap

Reviewer: ```suggestion
## Багаторядні гнучкі контейнери із flex-wrap
```

Хоч флексбокс і є одновимірною моделлю, можна змусити гнучкі елементи переходити на кілька рядів. При цьому слід розглядати кожен ряд як новий гнучкий контейнер. Будь-який розподіл простору діятиме для такого ряду окремо, без зв‘язку з іншими рядами з будь-якого боку.

Щоб ввімкнути переведення рядів, слід додати властивість {{cssxref("flex-wrap")}} зі значенням `wrap`. Після цього, якщо елементи завеликі, аби бути виведеними в одному ряду, вони будуть перенесені на наступний. Живий приклад нижче містить елементи зі заданою шириною, і сумарна ширина елементів завелика для гнучкого контейнера. Оскільки властивість `flex-wrap` має значення `wrap`, то елементи переходять на новий ряд. Якщо встановити значення `nowrap`, котре, до речі, є початковим, то елементи скоротяться, щоб вміститись у контейнері, тому що вони мають початкові значення флексбокса, що дає елементам змогу скорочуватись. Використання `nowrap` призвело б до переповнення, якби елементи не могли скорочуватись, або якби не могли скоротитись настільки, щоб поміститися.

{{EmbedGHLiveSample("css-examples/flexbox/basics/flex-wrap.html", '100%', 400)}}

> Довідайтесь більше про переведення гнучких елементів на новий ряд у настановах [Опанування переведення гнучких елементів на новий ряд](/uk/docs/Web/CSS/CSS_Flexible_Box_Layout/Mastering_Wrapping_of_Flex_Items).

Reviewer: ```suggestion
Довідайтесь більше про перенесення гнучких елементів на новий ряд у настановах [Опанування переведення гнучких елементів на новий ряд](/uk/docs/Web/CSS/CSS_Flexible_Box_Layout/Mastering_Wrapping_of_Flex_Items).
```

У живому прикладі нижче спробуйте змінити перше значення на одне з дозволених значень `flex-direction` - `row`, `row-reverse`, `column` чи `column-reverse`, а також змінити друге на `wrap` чи `nowrap`.

{{EmbedGHLiveSample("css-examples/flexbox/basics/flex-flow.html", '100%', 400)}}

## Властивості, що застосовуються до гнучких елементів

Для більшого контролю над гнучкими елементами можна націлюватись на них напряму. Це можливо за допомогою трьох властивостей:

- {{cssxref("flex-grow")}}
- {{cssxref("flex-shrink")}}
- {{cssxref("flex-basis")}}

У цьому огляді буде короткий позір на ці властивості, а повніше розуміння можна набути з настанов [Керування співвідношеннями гнучких елементів за головною віссю](/uk/docs/Web/CSS/CSS_Flexible_Box_Layout/Controlling_Ratios_of_Flex_Items_Along_the_Main_Ax).

Перед усвідомленням цих властивостей слід взяти до уваги концепцію **доступного простору**. При зміні значень цих гнучких властивостей – змінюється спосіб розподілу доступного простору між елементами. Ця концепція також важлива при шикуванні елементів.

> Якщо є три 100-піксельні елементи у контейнері, що має ширину 500 пікселів, то простір, необхідний для розкладання елементів, – 300 пікселів. Це залишає 200 пікселів доступного простору. Якщо не змінити початкові значення, то флексбокс віддасть цей простір останньому елементові.

Reviewer: ```suggestion
Якщо є три 100-піксельні елементи у контейнері, що має ширину 500 пікселів, то простір, необхідний для розкладання елементів, – 300 пікселів. Це залишає 200 пікселів доступного простору. Якщо не змінити початкові значення, то флексбокс залишить цей простір вільним за останнім елементом.
```

### Властивість flex-shrink

Коли властивість `flex-grow` займається додаванням простору за головною віссю, властивість `flex-shrink` керує її відбором. Якщо в контейнері немає достатнього простору для розкладання елементів, і `flex-shrink` має значення додатного цілого числа, то елемент стає меншим за `flex-basis`. Як і з `flex-grow`, різні значення можуть бути присвоєні, щоб один елемент скорочувався швидше за інших: елемент із більшим значенням `flex-shrink` скорочуватиметься швидше за братерні елементи із меншими значеннями.

Мінімальний розмір елемента береться до уваги при роботі з дієвим розміром скорочення, тобто `flex-shrink` потенційно може здаватись менш послідовним у поведінці за `flex-grow`. Для більш детального розгляду того, як цей алгоритм працює, дивіться статтю [Керування співвідношеннями гнучких елементів за головною віссю](/uk/docs/Web/CSS/CSS_Flexible_Box_Layout/Controlling_Ratios_of_Flex_Items_Along_the_Main_Ax).

> > **Зверніть увагу:** Ці значення `flex-grow` та `flex-shrink` – пропорції. Зазвичай якби усі елементи мали `flex: 1 1 200px`, а тоді хотіли, щоб один елемент зростав з подвійною швидкістю, то задали б йому `flex: 2 1 200px`. Втім, також за бажання можна було б використати `flex: 10 1 200px` чи `flex: 20 1 200px`.

Reviewer: ```suggestion
> **Зверніть увагу:** Ці значення: `flex-grow` та `flex-shrink` – пропорції. Зазвичай якби усі елементи мали `flex: 1 1 200px`, а тоді хотіли, щоб один елемент зростав з подвійною швидкістю, то задали б йому `flex: 2 1 200px`. Втім, також за бажання можна було б використати `flex: 10 1 200px` чи `flex: 20 1 200px`.
```

### Скорочені значення гнучких властивостей

Рідко можна побачити використання окремо властивостей `flex-grow`, `flex-shrink` і `flex-basis`; натомість їх поєднують у скорочення {{cssxref("flex")}}. Скорочення `flex` дає змогу встановити три значення у такому порядку: `flex-grow`, `flex-shrink`, `flex-basis`.

> Живий приклад нижче дає змогу спробувати різні значення скорочення `flex`; пам‘ятайте, що перше значення – `flex-grow`. Коли це значення – додатне, то це означає, що елемент може зростати. Друге значення – `flex-shrink`: із додатним значенням тут елементи можуть скорочуватись, втім, це може відбутись лише за умови, що сума розмірів за головною віссю утворює переповнення. Останнє значення – `flex-basis`; це значення, котре елементи використовують як свій базовий розмір, від котрого відбувається зростання та скорочення.

Reviewer: ```suggestion
Живий приклад нижче дає змогу спробувати різні значення скорочення `flex`; пам‘ятайте, що перше значення – `flex-grow`. Коли це значення – додатне, то це означає, що елемент може зростати. Друге значення – `flex-shrink`: із додатним значенням тут елементи можуть скорочуватись, втім, це може відбутись лише за умови, що сума розмірів елементів за головною віссю утворює переповнення. Останнє значення – `flex-basis`; це значення, котре елементи використовують як свій базовий розмір, від котрого відбувається зростання та скорочення.
```

Властивість {{cssxref("justify-content")}} використовується для шикування елементів на головній осі, напрям якої встановлений властивістю `flex-direction`. Початкове значення – `flex-start`, що змушує елементи стати в початок контейнера, але можна також встановити значення `flex-end`, аби виставити їх у кінці, або `center`, щоб виставити у центрі.

Також можна використати значення `space-between`, аби захопити увесь незайнятий після розкладання елементів простір, і поділити його порівну між елементами так, щоб між кожними двома суміжними елементами був однаковий проміжок. Щоб зліва та справа від кожного елемента був однаковий простір, потрібно встановити значення `space-around`. З `space-around` же елементи отримують по половинці простору з обох боків. Іще можна змусити елементи мати однаковий простір навколо себе за допомогою значення `space-evenly`. З `space-evenly` елементи мають цілий однаковий простір з обох кінців.

Спробуйте наступні значення `justify-content` у живому прикладі:

- `flex-start`
- `flex-end`
- `center`
- `space-around`
- `space-between`
- `space-evenly`

{{EmbedGHLiveSample("css-examples/flexbox/basics/justify-content.html", '100%', 380)}}

> У статті [Шикування елементів у гнучкому контейнері](/uk/docs/Web/CSS/CSS_Flexible_Box_Layout/Aligning_Items_in_a_Flex_Container) ці властивості досліджуються із більшим заглибленням, аби отримати краще розуміння їх роботи. Втім, вищенаведені прості приклади є корисними у більшості випадків.

Reviewer: ```suggestion
У статті [Шикування елементів у гнучкому контейнері](/uk/docs/Web/CSS/CSS_Flexible_Box_Layout/Aligning_Items_in_a_Flex_Container) ці властивості досліджуються із більшим заглибленням, аби отримати краще розуміння їх роботи. Втім, у більшості випадків будуть корисними і наведені вище прості приклади.
```

## Наступні кроки

> Після прочитання цієї статті має бути розуміння основних можливостей Флексбоксу. У наступній статті розглядається те, [як ця специфікація пов‘язана з іншими частинами CSS](/uk/docs/Web/CSS/CSS_Flexible_Box_Layout/Relationship_of_Flexbox_to_Other_Layout_Methods).

Reviewer: ```suggestion
Після прочитання цієї статті має сформуватись розуміння основних можливостей Флексбоксу. У наступній статті розглядається те, [як ця специфікація пов‘язана з іншими частинами CSS](/uk/docs/Web/CSS/CSS_Flexible_Box_Layout/Relationship_of_Flexbox_to_Other_Layout_Methods).
```
